app/controllers/ConfirmationPageController.scala

```diff
@@ -51,121 +51,120 @@ trait ConfirmationPageController extends ERSReturnBaseController with Authentica
   def confirmationPage() = AuthorisedForAsync() {
     implicit user =>
       implicit request =>
-        showConfirmationPage()(user, request, hc)
+          showConfirmationPage()(user, request, hc)
   }
 
   def showConfirmationPage()(implicit authContext: AuthContext, request: Request[AnyRef], hc: HeaderCarrier): Future[Result] = {
-    if (request.session.get(screenSchemeInfo).isDefined == false) Logger.error(s"Session doesn't contain scheme info: ${request.session}")
-    val schemeRef: String = cacheUtil.getSchemeRefFromScreenSchemeInfo(request.session.get(screenSchemeInfo))
-    val sessionBundelRef: String = request.session.get("bundelRef").getOrElse("")
-    val sessionDateTimeSubmitted: String = request.session.get("dateTimeSubmitted").getOrElse("")
-    if (sessionBundelRef == "") {
-      cacheUtil.fetch[ErsMetaData](CacheUtil.ersMetaData, schemeRef).flatMap { all =>
-        if (all.sapNumber.isDefined == false) Logger.error(s"Did cache util fail for scheme ${schemeRef} all.sapNumber is empty: ${all}")
-        ersConnector.connectToEtmpSummarySubmit(all.sapNumber.get, jsonParser.getSubmissionJson(all.schemeInfo.schemeRef, all.schemeInfo.schemeType, all.schemeInfo.taxYear, "EOY-RETURN")).flatMap { bundle =>
-          cacheUtil.getAllData(bundle, all).flatMap { alldata =>
-            if (alldata.isNilReturn == PageBuilder.OPTION_NIL_RETURN) {
-              saveAndSubmit(alldata, all, bundle)
-            } else {
-              cacheUtil.fetch[String](CacheUtil.VALIDATED_SHEEETS, schemeRef).flatMap { validatedSheets =>
-                ersConnector.checkForPresubmission(all.schemeInfo, validatedSheets).flatMap { checkResult =>
-                  checkResult.status match {
-                    case 200 => {
-                      Logger.info(s"Check for presubmission success with status ${checkResult.status}.")
-                      saveAndSubmit(alldata, all, bundle)
+    cacheUtil.fetch[RequestObject](cacheUtil.ersRequestObject).flatMap { requestObject =>
+      if (request.session.get(screenSchemeInfo).isEmpty) Logger.error(s"Session doesn't contain scheme info: ${request.session}")
+      val schemeRef: String = requestObject.getSchemeReference
+      val sessionBundelRef: String = request.session.get("bundelRef").getOrElse("")
```

Comment:
I'm not sure how widely these keys are used - can you have a look at replacing these with Strongly Typed values? Maybe extract it to a helper if you think the overhead helps.

Reply:
Also, you've got Bundel - is it meant to be Bundle?